bpo-38250: [Enum] single-bit flags are canonical

ethanfurman opened the pull request:
Flag members are now divided by one-bit verses multi-bit, with multi-bit
being treated as aliases.  Iterating over a flag only returns the
contained single-bit flags.

Iterating, `repr()`, and `str()` show members in definition order.

When constructing combined-member flags, any extra integer values are either discarded (`CONFORM`), turned into `int`s (`EJECT`) or treated as errors (`STRICT`).  Flag classes can specify which of those three behaviors is desired:

    >>> class Test(Flag, boundary=CONFORM):
    ...     ONE = 1
    ...     TWO = 2
    ...
    >>> Test(5)
    <Test.ONE: 1>

Besides the three above behaviors, there is also `KEEP`, which should not be used unless necessary -- for example, `_convert_` specifies `KEEP` as there are flag sets in the stdlib that are incomplete and/or inconsistent (e.g. `ssl.Options`).  `KEEP` will, as the name suggests, keep all bits; however, iterating over a flag with extra bits will only return the canonical flags contained, not the extra bits.

<!-- issue-number: [bpo-38250](https://bugs.python.org/issue38250) -->
https://bugs.python.org/issue38250
<!-- /issue-number -->

ethanfurman committed:
[Enum] fix Flag iteration, repr(), and str()

Flag members are now divided by one-bit verses multi-bit, with multi-bit
being treated as aliases.  Iterating over a flag only returns the
contained single-bit flags.

repr() and str() now only show the Flags, not extra integer values; any
extra integer values are either discarded (CONFORM), turned into
``int``s (EJECT) or treated as errors (STRICT).  Flag classes can
specify which of those three behaviors is desired:

    >>> class Test(Flag, boundary=CONFORM):
    ...     ONE = 1
    ...     TWO = 2
    ...
    >>> Test(5)
    <Test.ONE: 1>

ethanfurman committed:
add boundary KEEP for Flags (default for _convert_)

some flag sets, such as ``ssl.Options`` are incomplete/inconsistent;
using KEEP allows those flags to exist, and have useful repr()s, etc.

also, add ``_inverted_`` attribute to Flag members to significantly
speed up that operation.

ethanfurman committed:
update re.RegexFlag for new Flag implementation

repr() has been modified to support as closely as possible its previous
output; the big difference is that inverted flags cannot be output as
before because the inversion operation now always returns the comparable
positive result; i.e.

   re.A|re.I|re.M|re.S is ~(re.L|re.U|re.S|re.T|re.DEBUG)

in both of the above terms, the ``value`` is 282.

re's tests have been updated to reflect the modifications to repr().

ethanfurman committed:
remove extra white space

ethanfurman committed:
test that zero-valued members are empty

ethanfurman committed:
update tests to confirm CONFORM with negate

ethanfurman committed:
update aenum.rst; add doctest to test_enum

belm0 commented on Lib/enum.py:
```diff
@@ -56,6 +64,82 @@ def _is_private(cls_name, name):
     else:
         return False
 
+def _bits(num):
```
I couldn't figure out why the Flag implementation needs this string binary representation.  Why doesn't the code just work with binary operations on numbers?

I couldn't figure out why this custom string binary implementation is needed, rather than using `bin()`.

ethanfurman replied:
`bin()` does not properly display negative numbers:

    >>> bin(-10)
    '-0b1010'
    >>> _bits(-10)
    '0b10110'

The `_bits()` version actually displays which bits/flags are set.

belm0 replied:
_bits() is only used for error reporting-- the representation of negative numbers can be whatever we choose.

Anyway I think this works for `_bits()`:

```python3
def _bits(num):
    l = 2 ** (num).bit_length()
    return bin(num + l).replace('1', '0', 1) if num >= 0 \
        else bin(~num ^ (l - 1) + l)
```

ethanfurman replied:
Okay, but what does the `l` stand for?  I like to use real names so I have a prayer of understanding the function in the future.

belm0 replied:
`l` is "one above the most significant bit", somewhat hard to name

```python3
def _bin(num):
    """
    Like built-in bin(), except negative values are represented in twos-compliment, and the
    leading bit always indicates sign (0=positive, 1=negative).

    >>> _bin(10)
    '0b01010'
    >>> _bin(-10)
    '0b10110'
    """
    msb_placeholder = 2 ** (num).bit_length()
    return bin(num + msb_placeholder).replace('1', '0', 1) if num >= 0 \
        else bin(~num ^ (msb_placeholder - 1) + msb_placeholder)
```

ethanfurman replied:
`msb_placeholder` isn't much better for me, so I named it `ceiling` (which undoubtedly won't help somebody else).

Thanks for figuring that out!

belm0 commented on Lib/enum.py:
```diff
@@ -56,6 +64,82 @@ def _is_private(cls_name, name):
+        count += 1
+    return(count)
+
+def _bit_len(num):
```
I'm pretty sure this can be implemented in O(1)

ethanfurman replied:
Can PRs have PRs?  If so, I'm willing to look.  :-)

ethanfurman replied:
I don't know about O(1), but `int.bit_length()` is magnitudes faster.

belm0 commented on Lib/enum.py:
```diff
@@ -965,38 +1161,39 @@ def __contains__(self, other):
+                m
+                for m in sorted(
+                    self.__class__, key=lambda m: m._value_, reverse=True
+                    )
```
please consider:
  1. avoid the sort, by having the iteration of `__class__` be in the expected order.  (It's static and can be prepared in advance.). But see next suggestion.
  2. rather than iterate the Flag members, and compare each to the value, do the opposite: take each set bit in value and convert to a Flag member.  Because there may be 100 members in the enum, but only 1 bit set in the value.
  3. (stretch) change the defined order of `__iter__` (and of corresponding repr(), etc.) to be increasing value order, because that is what the set bit iteration algorithm can do efficiently.  Then the results don't need to be assembled into a list in advance, and the iter function can be a pure generator with no temporary collection.  It's significant: imagine again a flag with 100 members, and frequently iterating over values with many bits set.  It's creating a lot of garbage, and requiring upfront compute (wasteful if the iteration is stopped early).

ethanfurman replied:
1. The order of the class should be definition order (`list(Flag)`), but I find that a `repr()` in big-endian format is more useful (`list(Flag.member)`).
2. Would this support a member with unknown bits?
3. A generator option sounds interesting, and might be fine for `__iter__`, but I would still want `repr()` to be as-is (which should be fine -- if one is doing a `repr()` then speed shouldn't matter).

ethanfurman replied:
2. Done.

belm0 replied:
I don't agree about `repr()` or `str()` speed not being important.  Production systems rely on these for metrics reporting, logging, visualization, etc.

ethanfurman replied:
That's a good point.  However, while speed is important, it's not the most important, or we wouldn't have `Enum` to begin with.

I also see that iterating from msb to lsb is leaking an implementation detail, but so is lsb to msb.  Really, it should be iterating in member definition order, just like Enum does.

So what I'll do is:  if the `Flag` 's definition order is lsb to msb, then the more efficient iterator you proposed is used, otherwise a less efficient iterator that still honors definition order is used.

For the definition-order iterator I can either check every flag against the target, or use the lsb iterator and sort afterwards using an added `_def_order_` attribute.

ethanfurman replied:
Summarizing changes:

1. Iteration order is definition order.
2. Iteration is over bits in the member instead of bits in the class.
3. Member iteration is also in definition order, but if definition order is increasing value order then the more efficient lsb algorithm is used.

belm0 replied:
>  Really, it should be iterating in member definition order, just like Enum does.

I kind of followed why you wanted Flag class iter to be in definition order, following Enum.

But I don't understand why it's important to follow definition order when iterating the individual elements of a composite value.  Flag is a set (in my view), and a set doesn't have ordering-- so iterate in whatever order is efficient for the implementation.

> So what I'll do is: if the Flag 's definition order is lsb to msb, then the more efficient iterator you proposed is used, otherwise a less efficient iterator that still honors definition order is used.

It's unfortunately more complication for both the implementation and the user.

ethanfurman replied:
I see it as less complication for both: the "correct" iterator is chosen at class definition, then never changes; user's always get definition order.  How is that more complication?

belm0 commented on Lib/enum.py:
```diff
@@ -56,6 +64,82 @@ def _is_private(cls_name, name):
+    while(num):
+        num &= num - 1
+        count += 1
+    return(count)
```
remove parens in `while` and `return`

ethanfurman replied:
Oops.

belm0 commented on Lib/enum.py:
```diff
@@ -414,6 +510,33 @@ def __new__(metacls, cls, bases, classdict, **kwds):
+                i = 2**i
+                all_bit_total |= i
+                if i & multi_bit_total and not i & single_bit_total:
+                    missed.append(i)
```
not sure why this iteration is needed-- isn't the result of `multi_bit_total & ~single_bit_total` the "missed bits", and we can just do a bit walk of that value?

ethanfurman replied:
Changed.

belm0 commented on Lib/enum.py:
```diff
@@ -926,35 +1085,72 @@ def _generate_next_value_(name, start, count, last_values):
+            neg_value = value
+            value = cls._all_bits_ + 1 + value
+        # get members
+        members, unknown = _decompose(cls, value)
```
why is decompose still needed?

ethanfurman replied:
It does the work of figuring out which bits are valid flags.  I did switch it to walking the bits of `value`, and not worrying about negative numbers (since it is not called with any) and it is now about 4 times faster.

belm0 replied:
please consider:

  * remove the "unknown" calculation from `_decompose()`, since it's simply `value & cls._flag_mask_`
  * rename `_decompose()` to `_iter_members()` and make it a generator
  * then the instance `__iter__` implementation is just `return _iter_members(self, self._value_)`

ethanfurman replied:
- Done.
- Moved to class as `_iter_member_by_value_`, aka `_iter_member_`.
- `yield from self._iter_member_()`

Note that if definition order is not the same as lsb order, `_iter_member_by_def_` is used instead of `_iter_member_by_value_`.

belm0 commented on Lib/enum.py:
```diff
@@ -1017,86 +1214,60 @@ def __xor__(self, other):
+                        _or_,
+                        [m._value_ for m in self.__class__ if m not in self],
+                        0
+                        ))
```
it should just be a bit operation... like `flag_mask ^ self._value`

ethanfurman replied:
Change made.

belm0 reviewed (commented):
the documented semantic changes look good 👍 

CONFORM functionality looks useful, it would replace this hack I've got:
<details>
  <summary>SafeFlag</summary>

```python3
class SafeFlag(enum.Flag):
    """Replacement for enum.Flag which ignores unknown bits.

      * for Enum, unknown value should be an error
      * but for Flag, usually unknown bits can be safely ignored--
        this is a simple way to achieve backwards compatibility
      * use SafeFlag when flag values come from external input
        (another service, firmware, etc.)
    """

    @classmethod
    def _missing_(cls, value):
        # constrain to known bits
        # NOTE: enum subclass cannot define members, so using getattr
        all_value = getattr(cls, '_all_value', None)
        if all_value is None:
            all_value = sum(v.value for v in cls)
            setattr(cls, '_all_value', all_value)
        return super()._missing_(value & all_value)
```
</details>

Addition of the boundary options increases complexity however, and the implementation is growing rather than shrinking.

I'd like to help get the implementation into shape towards these ideals:
  * all operations should be O(1) or O(number of set bits).  Use bit operations rather than iteration wherever possible.
  * iterators should generate values on the fly rather than pre-assemble in memory (implies no sorting or reversing)
  * consider eliminating caching like `_value2member_map_`, assuming previous efficiency points can be achieved
  * eliminate _decompose 🙏 

I reviewed some of the enum.py changes, but there is a lot to cover.  Just raising the larger points to start.

ethanfurman committed:
fix doc test

ethanfurman committed:
optimizations

ethanfurman committed:
formatting

ethanfurman committed:
add news entry

ethanfurman committed:
fix formatting of news entry

belm0 commented on Lib/enum.py:
```diff
@@ -965,38 +1223,39 @@ def __contains__(self, other):
+            i = 2 ** (val.bit_length() - 1)
+            member = self._value2member_map_.get(i)
+            if member is not None:
+                yield member
```
I think this filter is not needed if value is masked with `_flag_mask_` upfront

ethanfurman replied:
Changed.

belm0 commented on Lib/enum.py:
```diff
@@ -539,42 +697,54 @@ def __setattr__(cls, name, value):
+                isinstance(names, (tuple, list))
+                and names
+                and isinstance(names[0], str)
+            ):
```
Would you consider making formatting changes in a separate PR?  There is a lot to carefully review here and these can be distracting.

ethanfurman replied:
Sorry, part of that is supposed to make it easier to read, and it's now a few commits back.  I don't have that much bandwidth available.  :(

belm0 replied:
there's also the bandwidth of all the people that will review this PR to consider-- and the PR should be set up to get good reviews out of them (https://mtlynch.io/code-review-love/#6-separate-functional-and-non-functional-changes)

about 25 of the edits in enum.py are formatting changes-- it would help a lot if they weren't mixed with the code changes that need careful review

ethanfurman replied:
You are, of course, correct.  In the future I will endeavor to keep them separate.

belm0 commented on Lib/enum.py:
```diff
@@ -414,6 +529,39 @@ def __new__(metacls, cls, bases, classdict, **kwds):
+                while missed_bits:
+                    i = 2 ** (missed_bits.bit_length() - 1)
+                    missed.append(i)
+                    missed_bits &= ~i
```
This pattern is repeated about 3 times, so it would be nice to have an `_iter_bits(val)` utility.

Then `missed` list is not needed and you can use the iterator within `join()` below

ethanfurman replied:
Done.

ethanfurman replied:
I also made the same change in `_missing_()`.

belm0 commented on Lib/enum.py:
```diff
@@ -414,6 +529,39 @@ def __new__(metacls, cls, bases, classdict, **kwds):
+                while missed_bits:
+                    i = 2 ** (missed_bits.bit_length() - 1)
+                    missed.append(i)
+                    missed_bits &= ~i
```
The other thing to say here, is that `bit_length()` is being used to walk the bits from MSB to LSB order.  I suspect it's significantly less efficient than the LSB to MSB walking of the Kernighan algorithm, because `bit_length()` probably has some math log calls, etc.  I'd like to try some `timeit` tests to verify.

This is why I'm suggesting that LSB to MSB be used in the public iterator, repr, etc.

belm0 replied:
walking from LSB to MSB is 2x faster

```python3
def bits_lsb(x):
    while x:
        b = x & (~x + 1)
        yield b
        x ^= b

def bits_msb(x):
    while x:
        b = 2 ** (x.bit_length() - 1)
        yield b
        x ^= b

>>> timeit.timeit('list(bits_lsb(65535))', globals=vars())
4.1443982159998995
>>> timeit.timeit('list(bits_msb(65535))', globals=vars())
9.664363892000438
```

ethanfurman replied:
Done.

hongweipeng commented on Lib/enum.py:
```diff
@@ -56,6 +64,73 @@ def _is_private(cls_name, name):
+        count += 1
+    return count
+
+def _is_single_bit(num):
```
> Since `int.bit_count()` was added in 3.10, I wonder if this would be more readable or give better performance.

ethanfurman replied:
It is slightly faster, but more importantly uses built-in functionality.  Change made.

ethanfurman committed:
update iteration method and order

Iteration is now in member definition order.  If member definition order
matches increasing value order, then a more efficient method of flag
decomposition is used; otherwise, sort() is called on the results of
that method to get definition order.

ethanfurman committed:
add John Belmonte

belm0 commented on Lib/enum.py:
```diff
@@ -414,6 +500,46 @@ def __new__(metacls, cls, bases, classdict, **kwds):
+                # while missed_bits:
+                #     i = 2 ** (missed_bits.bit_length() - 1)
+                #     missed.append(i)
+                #     missed_bits &= ~i
```
please delete

ethanfurman replied:
Just saw that.  Deleted.

belm0 commented on Lib/enum.py:
```diff
@@ -916,45 +1101,106 @@ def _generate_next_value_(name, start, count, last_values):
+        for val in _iter_bits_lsb(value):
+            member = cls._value2member_map_.get(val)
+            if member is not None:
+                yield member
```
can the None filtering be removed by masking the value?

```python3
for val in _iter_bits_lsb(value & cls._flag_mask_):
    yield cls._value2member_map_.get(val)
```

ethanfurman replied:
It can, and now is.

belm0 commented on Lib/enum.py:
```diff
@@ -71,13 +86,45 @@ def _break_on_call_reduce(self, proto):
+
+    ceiling = 2 ** (num).bit_length()
+    if num >= 0:
+        s = _bltin_bin(num + ceiling).replace('1', '0', 1)
```
why not leave the builtin `bin()` to its given name, and name the local version `_bin()`?

ethanfurman replied:
I like this `bin()` better.  It's not exported in `__all__`, and it's not unusual for modules to define their own versions of common operations without a leading `_`.  At some point in the future I may put it in `__all__`.

belm0 commented on Lib/enum.py:
```diff
@@ -71,13 +86,45 @@ def _break_on_call_reduce(self, proto):
+    twos-compliment, and the leading bit always indicates sign
+    (0=positive, 1=negative).
+
+    >>> bin(10)
+    '0b0 1010'
+    >>> bin(~10)   # ~10 is -11
+    '0b1 0101'
```
the space delimiter is quite strange-- I'm skeptical that it's going to help anybody

ethanfurman replied:
It keeps the sign bit separate from the flag bits.  It helps me, at least.  At some point I hope to have the flag portion in nibbles.

belm0 commented on Lib/enum.py:
```diff
@@ -414,6 +500,46 @@ def __new__(metacls, cls, bases, classdict, **kwds):
+            # set correct __iter__
+            inverted = ~enum_class(0)
+            if list(enum_class) != list(inverted):
+                if '|' in inverted._name_:
```
perhaps `!is_single_bit(inverted.value)` is more explicit

ethanfurman replied:
I don't think that would work -- I'm trying to discern between a brand new combination of all flags, versus an in-the-class alias of all flags -- `WHITE` in `Color`, for example.

ethanfurman replied:
Noise removed.  I'd like to get this committed by the weekend.

belm0 reviewed (commented):
it's a complicated change to a complicated implementation (and sprinkled with noise)... I need to try another review pass later

ethanfurman committed:
more bit-fiddling improvements

ethanfurman committed:
use pop() instead of "del"

new composite members aren't always added to _value2member_map_ -- this
ensures the operation succeeds

ethanfurman committed:
update DynamicClassAttribute __doc__

ethanfurman committed:
remove formatting changes

ethanfurman committed:
remove extra parens

ethanfurman committed:
remove formatting changes

also disable doctests until formatting changes are added back

ethanfurman committed:
remove formatting

ethanfurman committed:
simplify determination of member iteration

belm0 commented on Lib/enum.py:
```diff
@@ -153,15 +199,29 @@ def __set_name__(self, enum_class, member_name):
         enum_member._name_ = member_name
         enum_member.__objclass__ = enum_class
         enum_member.__init__(*args)
+        enum_member._sort_order_ = len(enum_class._member_names_)
```
this doesn't look right-- `_sort_order_` is an int?

ethanfurman replied:
Yes, because sorting `int`s is easy.

belm0 replied:
This code through me off because it's not clear how `len(enum_class._member_names_)` provides the sort order, so it may warrant a comment.

belm0 commented on Lib/enum.py:
```diff
@@ -414,6 +484,40 @@ def __new__(metacls, cls, bases, classdict, **kwds):
+                    single_bit_total |= flag._value_
+                else:
+                    # multi-bit flags are considered aliases
+                    multi_bit_total |= flag._value_
```
it would be better to put `flag._value_` in a local var rather than 3 dereferencing expressions

ethanfurman replied:
Done.

belm0 replied:
did you forget to push some pending commits?

ethanfurman replied:
Not forgot, just not gotten to it.  Hope to get it done and committed today (Monday).

belm0 commented on Lib/enum.py:
```diff
@@ -916,45 +1042,104 @@ def _generate_next_value_(name, start, count, last_values):
+        members = list(cls._iter_member_by_value_(value))
+        members.sort(key=lambda m: m._sort_order_)
+        for member in members:
+            yield member
```
```python3
yield from sorted(cls._iter_member_by_value_(value), key=lambda m: m._sort_order_)
```

ethanfurman replied:
Done.

ethanfurman committed:
add note about next auto() value for Enum and Flag

ethanfurman committed:
local name optimizations

ethanfurman committed:
remove commented-out code

ethanfurman committed:
add test for next auto() and _order_

auto() for flags returns the first power of two not used

_order_ has any names that are aliases removed before checking against
_member_names_

ethanfurman committed:
raise TypeError if _value_ not added in custom new

also fix _order_ tests

ethanfurman committed:
enable doc tests, update formatting

ethanfurman committed:
fix note

ethanfurman committed:
Update 2021-01-14-15-07-16.bpo-38250.1fvhOk.rst
